src/utils.cpp

```diff
@@ -521,13 +575,101 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
     candidate_paths = ::expand_wildcard_paths(candidate_paths);
 
   std::vector<std::string> valid_executable_paths;
-  for (const auto& path : candidate_paths)
-    if (access(path.c_str(), X_OK) == 0)
-      valid_executable_paths.push_back(path);
+  for (const auto &path : candidate_paths)
+  {
+    std::string rel_path;
+    if (pid > 0 && pid_in_different_mountns(pid))
+      rel_path = path_for_pid_mountns(pid, path);
+    else
+      rel_path = path;
+    if (access(rel_path.c_str(), X_OK) == 0)
+      valid_executable_paths.push_back(rel_path);
+  }
 
   return valid_executable_paths;
 }
 
+std::string path_for_pid_mountns(int pid, const std::string &path)
+{
+  std::ostringstream pid_relative_path;
+  std::string root =
+      (path.length() >= 1 && path.at(0) == '/') ? "/root" : "/root/";
+  pid_relative_path << "/proc/" << pid << root << path;
+  return pid_relative_path.str();
+}
+
+/*
+Determines if the target process is in a different mount namespace from
+bpftrace.
+
+If a process is in a different mount namespace (eg, container) it is very
+likely that any references to local paths will not be valid, and that paths
+need to be made relative to the PID.
+
+If an invalid PID is specified or doesn't exist, it returns false.
+True is only returned if the namespace of the target process could be read and
+it doesn't match that of bpftrace.
+*/
+static bool pid_in_different_mountns(int pid)
+{
+
+  struct stat self_stat, target_stat;
+  int self_fd = -1, target_fd = -1;
+  std::string errmsg;
+  char buf[64];
+
+  if (pid <= 0)
+    return false;
+
+  if ((size_t)snprintf(buf, sizeof(buf), "/proc/%d/ns/mnt", pid) >= sizeof(buf))
+  {
+    errmsg = "Reading mountNS would overflow buffer.";
+    goto error;
+  }
+
+  self_fd = open("/proc/self/ns/mnt", O_RDONLY);
+  if (self_fd < 0)
+  {
+    errmsg = "open(/proc/self/ns/mnt)";
+    goto error;
+  }
+
+  target_fd = open(buf, O_RDONLY);
+  if (target_fd < 0)
+  {
+    errmsg = "open(/proc/<pid>/ns/mnt)";
+    goto error;
+  }
+
+  if (fstat(self_fd, &self_stat))
+  {
+    errmsg = "fstat(self_fd)";
+    goto error;
+  }
+
+  if (fstat(target_fd, &target_stat))
+  {
+    errmsg = "fstat(target_fd)";
```

Comment:
Knowing why it failed can be useful

```suggestion
    errmsg << "fstat(target_fd): " << strerror(errno);
```

(need to make errmsg a stringstream for this)

Reply:
Added thanks for the suggestion